### PR TITLE
#9755: move ttnn.concat to match the new file structure

### PR DESCRIPTION
**Pull request by sjameelTT**
- TODO: move program factory, ttlib C++ implementation, ttlib C++/python references and delete tt_eager implementation

### Ticket
#9755 

### Problem description
ttnn::concat needs to be restructured to match the new op format

### What's changed
- Concat logic moved into a concat_pybind.hpp and concat.hpp which defines the new struct
- TODO: move all the tt_eager program creation and kernels into the new folders and delete the tt_eager implementation and ttlib pybinding

### Checklist
- [x] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        // [dim](const ttnn::Tensor& input_tensor){
+        //    return input_tensor.get_layout() == ttnn::TILE_LAYOUT and not has_tile_padding(input_tensor, dim);
+        //});
+        // TT_FATAL(all_tensors_are_tile_layout_without_padding, "Not Implemented");
+
+        const ttnn::Tensor& first_tensor = input_tensors.front();
+        const int rank = first_tensor.get_shape().rank();
+
+        // Wrap dim
+        dim = dim < 0 ? rank + dim : dim;
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp:41]: I think we have some `normalize dim` function to help with this?

**Review by TT-BrianLiu — APPROVED** [review]
Minor comment

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat_pybind.hpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                const int dim,
+                std::optional<ttnn::Tensor> optional_output_tensors,
+                std::optional<ttnn::MemoryConfig>& memory_config,
+                uint8_t queue_id) {
+                    return self(queue_id, tensors, dim, memory_config);
+                },
+                py::arg("tensors"),
+                py::arg("dim") = 0,
+                py::kw_only(),
+                py::arg("out") = std::nullopt,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/concat/concat_pybind.hpp:59]: I think we call it `output_tensor` usually. Lets keep consistent

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp (hunk 1 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat.hpp]: Yep, switched to that.

File: ttnn/cpp/ttnn/operations/data_movement/concat/concat_pybind.hpp (hunk 2 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/concat/concat_pybind.hpp]: Switched to `output_tensor`

**Pull request merged**